File: code/packages/python/directed-graph/tests/test_graph.py
"""
test_graph.py -- Tests for DirectedGraph Core Operations
=========================================================

These tests cover the fundamental data structure operations of the directed
graph: adding/removing nodes and edges, querying neighbors, and error handling.
We leave algorithm tests (topological sort, cycle detection, etc.) to
``test_algorithms.py`` so each file has a clear focus.

The tests are organized from simplest to most complex, following a natural
progression that mirrors how you'd learn the API:

1. Empty graph behavior
2. Single node operations
3. Single edge operations
4. Multi-node/edge operations
5. Error conditions
6. Edge cases (duplicates, repr, etc.)

Each test function is named to describe WHAT it tests and WHAT the expected
outcome is, so you can read the test names as a specification of the graph's
behavior.
"""

import pytest

from directed_graph import (
    DirectedGraph,
    EdgeNotFoundError,
    NodeNotFoundError,
)


# ======================================================================
# 1. Empty Graph
# ======================================================================
# An empty graph has no nodes and no edges. All query methods should return
# empty collections rather than raising errors.


class TestEmptyGraph:
    """Verify that a freshly created graph behaves correctly."""

    def test_empty_graph_has_no_nodes(self) -> None:
        """An empty graph should report zero nodes."""
        g = DirectedGraph()
        assert g.nodes() == []
        assert len(g) == 0

    def test_empty_graph_has_no_edges(self) -> None:
        """An empty graph should report zero edges."""
        g = DirectedGraph()
        assert g.edges() == []

    def test_empty_graph_has_node_returns_false(self) -> None:
        """has_node should return False for any node in an empty graph."""
        g = DirectedGraph()
        assert g.has_node("A") is False
        assert "A" not in g

    def test_empty_graph_has_edge_returns_false(self) -> None:
        """has_edge should return False for any edge in an empty graph."""
        g = DirectedGraph()
        assert g.has_edge("A", "B") is False


# ======================================================================
# 2. Single Node
# ======================================================================
# Adding a single node is the simplest mutation. We test that it appears
# in the graph and can be removed.


class TestSingleNode:
    """Test adding, querying, and removing a single node."""

    def test_add_node_makes_it_present(self) -> None:
        """After add_node, the node should be findable."""
        g = DirectedGraph()
        g.add_node("A")
        assert g.has_node("A") is True
        assert "A" in g
        assert len(g) == 1
        assert g.nodes() == ["A"]

    def test_remove_node_makes_it_absent(self) -> None:
        """After remove_node, the node should be gone."""
        g = DirectedGraph()
        g.add_node("A")
        g.remove_node("A")
        assert g.has_node("A") is False
        assert len(g) == 0

    def test_add_node_is_idempotent(self) -> None:
        """Adding the same node twice should not create duplicates."""
        g = DirectedGraph()
        g.add_node("A")
        g.add_node("A")  # Should be a no-op
        assert len(g) == 1
        assert g.nodes() == ["A"]

    def test_predecessors_of_isolated_node(self) -> None:
        """An isolated node has no predecessors."""
        g = DirectedGraph()
        g.add_node("A")
        assert g.predecessors("A") == []

    def test_successors_of_isolated_node(self) -> None:
        """An isolated node has no successors."""
        g = DirectedGraph()
        g.add_node("A")
        assert g.successors("A") == []


# ======================================================================
# 3. Single Edge
# ======================================================================
# An edge implicitly adds both endpoints, so add_edge("A", "B") creates
# nodes A and B plus the edge A -> B.


class TestSingleEdge:
    """Test adding a single edge and querying its properties."""

    def test_add_edge_creates_both_nodes(self) -> None:
        """add_edge should implicitly add both endpoint nodes."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        assert g.has_node("A")
        assert g.has_node("B")
        assert len(g) == 2

    def test_add_edge_creates_the_edge(self) -> None:
        """After add_edge, has_edge should return True."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        assert g.has_edge("A", "B") is True
        # The reverse direction should NOT exist.
        assert g.has_edge("B", "A") is False

    def test_edges_returns_the_edge(self) -> None:
        """edges() should include the added edge."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        assert g.edges() == [("A", "B")]

    def test_predecessors_and_successors(self) -> None:
        """For edge A -> B: A's successor is B, B's predecessor is A."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        assert g.successors("A") == ["B"]
        assert g.predecessors("B") == ["A"]
        # A has no predecessors, B has no successors.
        assert g.predecessors("A") == []
        assert g.successors("B") == []

    def test_remove_edge(self) -> None:
        """remove_edge should delete the edge but keep both nodes."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        g.remove_edge("A", "B")
        assert g.has_edge("A", "B") is False
        # Both nodes should still exist.
        assert g.has_node("A")
        assert g.has_node("B")

    def test_duplicate_edge_is_idempotent(self) -> None:
        """Adding the same edge twice should not create duplicates."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        g.add_edge("A", "B")  # Should be a no-op
        assert g.edges() == [("A", "B")]


# ======================================================================
# 4. Multi-Node Operations
# ======================================================================
# Test more complex graph structures to make sure the adjacency dicts
# stay consistent.


class TestMultiNodeOperations:
    """Test graphs with multiple nodes and edges."""

    def test_remove_node_cleans_up_edges(self) -> None:
        """Removing a node should remove all its incoming and outgoing edges.

        Given A -> B -> C, removing B should leave A and C as isolated nodes
        with no edges between them.
        """
        g = DirectedGraph()
        g.add_edge("A", "B")
        g.add_edge("B", "C")
        g.remove_node("B")

        assert not g.has_node("B")
        assert g.has_node("A")
        assert g.has_node("C")
        assert g.edges() == []
        assert g.successors("A") == []
        assert g.predecessors("C") == []

    def test_remove_node_with_multiple_edges(self) -> None:
        """Removing a hub node should clean up all connected edges.

        Given A -> B, C -> B, B -> D, removing B should leave A, C, D
        with no edges.
        """
        g = DirectedGraph()
        g.add_edge("A", "B")
        g.add_edge("C", "B")
        g.add_edge("B", "D")
        g.remove_node("B")

        assert len(g) == 3
        assert g.edges() == []


# ======================================================================
# 5. Error Conditions
# ======================================================================
# These tests verify that the graph raises the right exceptions for
# invalid operations.


class TestErrorConditions:
    """Test that invalid operations raise appropriate errors."""

    def test_self_loop_raises_value_error(self) -> None:
        """A self-loop (A -> A) should raise ValueError."""
        g = DirectedGraph()
        with pytest.raises(ValueError, match="Self-loops are not allowed"):
            g.add_edge("A", "A")

    def test_remove_nonexistent_node_raises(self) -> None:
        """Removing a node that doesn't exist should raise NodeNotFoundError."""
        g = DirectedGraph()
        with pytest.raises(NodeNotFoundError):
            g.remove_node("X")

    def test_node_not_found_error_has_node_attribute(self) -> None:
        """NodeNotFoundError should carry the missing node value."""
        g = DirectedGraph()
        with pytest.raises(NodeNotFoundError) as exc_info:
            g.remove_node("X")
        assert exc_info.value.node == "X"

    def test_remove_nonexistent_edge_raises(self) -> None:
        """Removing an edge that doesn't exist should raise EdgeNotFoundError."""
        g = DirectedGraph()
        g.add_node("A")
        g.add_node("B")
        with pytest.raises(EdgeNotFoundError):
            g.remove_edge("A", "B")

    def test_edge_not_found_error_has_attributes(self) -> None:
        """EdgeNotFoundError should carry both node values."""
        g = DirectedGraph()
        with pytest.raises(EdgeNotFoundError) as exc_info:
            g.remove_edge("X", "Y")
        assert exc_info.value.from_node == "X"
        assert exc_info.value.to_node == "Y"

    def test_predecessors_of_nonexistent_node_raises(self) -> None:
        """predecessors() should raise NodeNotFoundError for missing nodes."""
        g = DirectedGraph()
        with pytest.raises(NodeNotFoundError):
            g.predecessors("X")

    def test_successors_of_nonexistent_node_raises(self) -> None:
        """successors() should raise NodeNotFoundError for missing nodes."""
        g = DirectedGraph()
        with pytest.raises(NodeNotFoundError):
            g.successors("X")


# ======================================================================
# 6. Edge Cases and Repr
# ======================================================================


class TestEdgeCases:
    """Test repr, contains, and other edge cases."""

    def test_repr_shows_counts(self) -> None:
        """__repr__ should show node and edge counts."""
        g = DirectedGraph()
        g.add_edge("A", "B")
        assert repr(g) == "DirectedGraph(nodes=2, edges=1)"

    def test_contains_is_same_as_has_node(self) -> None:
        """The 'in' operator should work like has_node."""
        g = DirectedGraph()
        g.add_node("A")
        assert ("A" in g) is True
        assert ("B" in g) is False

    def test_integer_nodes(self) -> None:
        """The graph should work with integer nodes, not just strings."""
        g = DirectedGraph()
        g.add_edge(1, 2)
        g.add_edge(2, 3)
        assert g.has_edge(1, 2)
        assert g.successors(1) == [2]
        assert len(g) == 3


# ======================================================================
# 7. Self-Loop Support
# ======================================================================
# When ``allow_self_loops=True``, edges from a node to itself are permitted.
# A self-loop means the node is both a successor AND predecessor of itself.
# This naturally creates a cycle, so cycle-related algorithms detect it.
#
# Why support self-loops? Some graph models require them:
# - State machines where a state can transition to itself
# - Dependency graphs where a package depends on its own build output
# - Labeled graphs where multiple relationships exist between entities


class TestSelfLoopDefault:
    """Verify that self-loops are rejected by default (backward compatible)."""

    def test_default_rejects_self_loop(self) -> None:
        """Without allow_self_loops, A -> A should raise ValueError."""
        g = DirectedGraph()
        with pytest.raises(ValueError, match="Self-loops are not allowed"):
            g.add_edge("A", "A")

    def test_default_rejects_integer_self_loop(self) -> None:
        """Self-loop rejection works for non-string node types too."""
        g = DirectedGraph()
        with pytest.raises(ValueError):
            g.add_edge(1, 1)


class TestSelfLoopAllowed:
    """Test graphs where allow_self_loops=True."""

    def test_self_loop_can_be_added(self) -> None:
        """With allow_self_loops=True, A -> A should succeed."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert g.has_edge("A", "A")

    def test_self_loop_node_is_own_successor(self) -> None:
        """A self-loop means the node appears in its own successor set."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert "A" in g.successors("A")

    def test_self_loop_node_is_own_predecessor(self) -> None:
        """A self-loop means the node appears in its own predecessor set."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert "A" in g.predecessors("A")

    def test_self_loop_appears_in_edges(self) -> None:
        """edges() should include the self-loop as a (A, A) tuple."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert ("A", "A") in g.edges()

    def test_self_loop_creates_cycle(self) -> None:
        """A graph with a self-loop has a cycle."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert g.has_cycle() is True

    def test_self_loop_plus_normal_edges(self) -> None:
        """Self-loops can coexist with normal edges."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "B")
        g.add_edge("B", "B")  # self-loop on B
        assert g.has_edge("A", "B")
        assert g.has_edge("B", "B")
        assert not g.has_edge("A", "A")

    def test_self_loop_transitive_closure(self) -> None:
        """Transitive closure from a self-loop node includes itself.

        If A -> A and A -> B, then transitive_closure("A") = {A, B}
        because A can reach itself via the self-loop.
        """
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        g.add_edge("A", "B")
        closure = g.transitive_closure("A")
        assert "A" in closure
        assert "B" in closure

    def test_self_loop_only_transitive_closure(self) -> None:
        """Transitive closure of a node with ONLY a self-loop includes itself."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert g.transitive_closure("A") == {"A"}

    def test_remove_node_cleans_up_self_loop(self) -> None:
        """Removing a node with a self-loop should not leave stale data."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        g.add_edge("A", "B")
        g.remove_node("A")
        assert not g.has_node("A")
        assert g.has_node("B")
        assert g.edges() == []

    def test_remove_edge_on_self_loop(self) -> None:
        """remove_edge should work on self-loop edges."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        g.remove_edge("A", "A")
        assert not g.has_edge("A", "A")
        assert g.has_node("A")  # node still exists

    def test_self_loop_topological_sort_raises(self) -> None:
        """Topological sort should raise CycleError for self-loop graphs."""
        from directed_graph import CycleError

        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        with pytest.raises(CycleError):
            g.topological_sort()

    def test_repr_shows_self_loops_flag(self) -> None:
        """repr should indicate allow_self_loops when True."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        assert "allow_self_loops=True" in repr(g)

    def test_self_loop_idempotent(self) -> None:
        """Adding the same self-loop twice is a no-op."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        g.add_edge("A", "A")
        assert g.edges() == [("A", "A")]

    def test_multiple_nodes_with_self_loops(self) -> None:
        """Multiple nodes can each have self-loops."""
        g = DirectedGraph(allow_self_loops=True)
        g.add_edge("A", "A")
        g.add_edge("B", "B")
        g.add_edge("A", "B")
        assert g.has_edge("A", "A")
        assert g.has_edge("B", "B")
        assert g.has_edge("A", "B")
        assert len(g.edges()) == 3

File: code/packages/ruby/directed_graph/test/test_graph.rb
# frozen_string_literal: true

require_relative "test_helper"

# --------------------------------------------------------------------------
# test_graph.rb — Tests for the core Graph data structure
# --------------------------------------------------------------------------
#
# We organise tests into logical groups:
#   1. Empty graph behaviour
#   2. Single-node operations
#   3. Adding and removing edges
#   4. Error conditions (self-loops, missing nodes/edges)
#   5. Predecessors and successors
#   6. Size and membership queries
# --------------------------------------------------------------------------

module CodingAdventures
  module DirectedGraph
    class TestGraphCore < Minitest::Test
      # ----------------------------------------------------------------
      # Empty graph
      # ----------------------------------------------------------------

      def test_empty_graph_has_no_nodes
        g = Graph.new
        assert_equal [], g.nodes
      end

      def test_empty_graph_has_no_edges
        g = Graph.new
        assert_equal [], g.edges
      end

      def test_empty_graph_size_is_zero
        g = Graph.new
        assert_equal 0, g.size
      end

      # ----------------------------------------------------------------
      # Single node
      # ----------------------------------------------------------------

      def test_add_single_node
        g = Graph.new
        g.add_node("A")
        assert_equal ["A"], g.nodes
        assert_equal 1, g.size
      end

      def test_add_duplicate_node_is_noop
        g = Graph.new
        g.add_node("A")
        g.add_node("A")
        assert_equal ["A"], g.nodes
        assert_equal 1, g.size
      end

      def test_add_node_returns_self_for_chaining
        g = Graph.new
        result = g.add_node("A")
        assert_same g, result
      end

      def test_has_node_returns_true_for_existing
        g = Graph.new
        g.add_node("X")
        assert g.has_node?("X")
      end

      def test_has_node_returns_false_for_missing
        g = Graph.new
        refute g.has_node?("X")
      end

      # ----------------------------------------------------------------
      # Adding edges
      # ----------------------------------------------------------------

      def test_add_edge_creates_both_nodes
        g = Graph.new
        g.add_edge("A", "B")
        assert g.has_node?("A")
        assert g.has_node?("B")
        assert_equal 2, g.size
      end

      def test_add_edge_creates_directed_connection
        g = Graph.new
        g.add_edge("A", "B")
        assert g.has_edge?("A", "B")
        refute g.has_edge?("B", "A")
      end

      def test_add_duplicate_edge_is_noop
        g = Graph.new
        g.add_edge("A", "B")
        g.add_edge("A", "B")
        assert_equal [["A", "B"]], g.edges
      end

      def test_add_edge_returns_self_for_chaining
        g = Graph.new
        result = g.add_edge("A", "B")
        assert_same g, result
      end

      def test_self_loop_raises_cycle_error
        g = Graph.new
        assert_raises(CycleError) { g.add_edge("A", "A") }
      end

      def test_self_loop_error_message_includes_node
        g = Graph.new
        error = assert_raises(CycleError) { g.add_edge("A", "A") }
        assert_match(/Self-loop/, error.message)
        assert_match(/"A"/, error.message)
      end

      # ----------------------------------------------------------------
      # Removing nodes
      # ----------------------------------------------------------------

      def test_remove_node_deletes_node
        g = Graph.new
        g.add_node("A")
        g.remove_node("A")
        refute g.has_node?("A")
        assert_equal 0, g.size
      end

      def test_remove_node_cleans_up_outgoing_edges
        g = Graph.new
        g.add_edge("A", "B")
        g.remove_node("A")
        refute g.has_edge?("A", "B")
        assert_equal [], g.predecessors("B")
      end

      def test_remove_node_cleans_up_incoming_edges
        g = Graph.new
        g.add_edge("A", "B")
        g.remove_node("B")
        refute g.has_edge?("A", "B")
        assert_equal [], g.successors("A")
      end

      def test_remove_missing_node_raises_error
        g = Graph.new
        assert_raises(NodeNotFoundError) { g.remove_node("Z") }
      end

      def test_remove_node_returns_self_for_chaining
        g = Graph.new
        g.add_node("A")
        result = g.remove_node("A")
        assert_same g, result
      end

      # ----------------------------------------------------------------
      # Removing edges
      # ----------------------------------------------------------------

      def test_remove_edge
        g = Graph.new
        g.add_edge("A", "B")
        g.remove_edge("A", "B")
        refute g.has_edge?("A", "B")
        # Both nodes should still exist.
        assert g.has_node?("A")
        assert g.has_node?("B")
      end

      def test_remove_edge_missing_source_raises
        g = Graph.new
        g.add_node("B")
        assert_raises(NodeNotFoundError) { g.remove_edge("A", "B") }
      end

      def test_remove_edge_missing_target_raises
        g = Graph.new
        g.add_node("A")
        assert_raises(NodeNotFoundError) { g.remove_edge("A", "B") }
      end

      def test_remove_nonexistent_edge_raises
        g = Graph.new
        g.add_node("A")
        g.add_node("B")
        assert_raises(EdgeNotFoundError) { g.remove_edge("A", "B") }
      end

      def test_remove_edge_returns_self_for_chaining
        g = Graph.new
        g.add_edge("A", "B")
        result = g.remove_edge("A", "B")
        assert_same g, result
      end

      # ----------------------------------------------------------------
      # Predecessors and successors
      # ----------------------------------------------------------------

      def test_predecessors
        g = Graph.new
        g.add_edge("A", "C")
        g.add_edge("B", "C")
        assert_equal %w[A B], g.predecessors("C")
      end

      def test_successors
        g = Graph.new
        g.add_edge("A", "B")
        g.add_edge("A", "C")
        assert_equal %w[B C], g.successors("A")
      end

      def test_predecessors_missing_node_raises
        g = Graph.new
        assert_raises(NodeNotFoundError) { g.predecessors("Z") }
      end

      def test_successors_missing_node_raises
        g = Graph.new
        assert_raises(NodeNotFoundError) { g.successors("Z") }
      end

      def test_predecessors_of_root_is_empty
        g = Graph.new
        g.add_edge("A", "B")
        assert_equal [], g.predecessors("A")
      end

      def test_successors_of_leaf_is_empty
        g = Graph.new
        g.add_edge("A", "B")
        assert_equal [], g.successors("B")
      end

      # ----------------------------------------------------------------
      # Edges listing
      # ----------------------------------------------------------------

      def test_edges_returns_all_pairs_sorted
        g = Graph.new
        g.add_edge("B", "C")
        g.add_edge("A", "B")
        assert_equal [["A", "B"], ["B", "C"]], g.edges
      end

      # ----------------------------------------------------------------
      # has_edge? on missing nodes
      # ----------------------------------------------------------------

      def test_has_edge_returns_false_for_missing_source
        g = Graph.new
        g.add_node("B")
        refute g.has_edge?("A", "B")
      end

      def test_has_edge_returns_false_for_missing_target
        g = Graph.new
        g.add_node("A")
        refute g.has_edge?("A", "B")
      end

      # ----------------------------------------------------------------
      # allow_self_loops flag
      # ----------------------------------------------------------------
      #
      # The allow_self_loops flag controls whether edges from a node to
      # itself are permitted.  By default they are rejected with a
      # CycleError, but when enabled they behave like any other edge.

      def test_default_disallows_self_loops
        g = Graph.new
        refute g.allow_self_loops?
      end

      def test_allow_self_loops_flag_true
        g = Graph.new(allow_self_loops: true)
        assert g.allow_self_loops?
      end

      def test_self_loop_allowed_when_flag_true
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        assert g.has_edge?("A", "A")
      end

      def test_self_loop_node_is_own_successor
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        assert_includes g.successors("A"), "A"
      end

      def test_self_loop_node_is_own_predecessor
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        assert_includes g.predecessors("A"), "A"
      end

      def test_self_loop_appears_in_edges
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        assert_includes g.edges, ["A", "A"]
      end

      def test_self_loop_size_is_one
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        assert_equal 1, g.size
      end

      def test_self_loop_with_other_edges
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        g.add_edge("A", "B")
        assert g.has_edge?("A", "A")
        assert g.has_edge?("A", "B")
        assert_equal 2, g.size
      end

      def test_remove_self_loop
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        g.remove_edge("A", "A")
        refute g.has_edge?("A", "A")
        assert g.has_node?("A")
      end

      def test_remove_node_with_self_loop
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        g.add_edge("A", "B")
        g.remove_node("A")
        refute g.has_node?("A")
        assert g.has_node?("B")
        assert_equal [], g.edges
      end

      def test_self_loop_duplicate_is_noop
        g = Graph.new(allow_self_loops: true)
        g.add_edge("A", "A")
        g.add_edge("A", "A")
        assert_equal [["A", "A"]], g.edges
      end

      def test_self_loop_still_raises_when_flag_false
        g = Graph.new(allow_self_loops: false)
        assert_raises(CycleError) { g.add_edge("A", "A") }
      end
    end
  end
end
